src/Workspaces/Remote/ServiceHub/Services/SemanticClassification/RemoteSemanticClassificationService.Caching.cs

```diff
@@ -93,22 +93,28 @@ private static string GetPersistenceName(ClassificationType type)
             ImmutableSegmentedList<(Document document, ClassificationType type, ClassificationOptions options)> documents,
             CancellationToken cancellationToken)
         {
-            // Group all the requests by document (as we may have gotten many requests for the same document). Then,
-            // only process the last document from each group (we don't need to bother stale versions of a particular
-            // document).
-            var groups = documents.GroupBy(d => d.document.Id);
-            var tasks = groups.Select(g => Task.Run(() =>
+            using var _ = ArrayBuilder<Task>.GetInstance(out var tasks);
+
+            // First group by type.  That way we process the last semantic and last embedded-lang classifications per document.
+            foreach (var typeGroup in documents.GroupBy(t => t.type))
             {
-                var (document, type, options) = g.Last();
-                return CacheClassificationsAsync(document, type, options, cancellationToken);
-            }, cancellationToken));
+                // Then, group all those requests by document (as we may have gotten many requests for the same
+                // document). Then, only process the last document from each group (we don't need to bother stale
+                // versions of a particular document).
+                foreach (var group in typeGroup.GroupBy(d => d.document.Id))
+                {
+                    var (document, type, options) = group.Last();
+                    tasks.Add(CacheClassificationsAsync(document, type, options, cancellationToken));
+                }
+            }
 
             await Task.WhenAll(tasks).ConfigureAwait(false);
         }
 
         private static async Task CacheClassificationsAsync(
             Document document, ClassificationType type, ClassificationOptions options, CancellationToken cancellationToken)
         {
+            await Task.Yield();
```

Comment:
Where is the heavy part of this method that we're trying to prevent from running synchronously from the caller? If it's really cheap to get through the ChecksumMatchesAsync call and it commonly indicates that they match, would it make sense to move this to after that point?

Reply:
Actually, i'm fine just making this serial, and not doing parallelism here.  Practically, this should just be a few docs, and this will help ensure that he BG is not taking up too many resources.